agent/lib/kontena/workers/service_pod_worker.rb

```diff
@@ -62,17 +67,63 @@ def on_container_event(topic, e)
       if attrs['io.kontena.service.id'] == @service_pod.service_id &&
           attrs['io.kontena.service.instance_number'].to_i == @service_pod.instance_number
         @container_state_changed = true
+        handle_restart_on_die if DIE_EVENTS.include?(e.status)
+      end
+    end
+
+    # Handles events when container has died
+    def handle_restart_on_die
+      unless @service_pod.running?
+        @restart_backoff_timer.cancel if @restart_backoff_timer
+        return
+      end
+
+      return if apply_in_progress?
+
+      # cancel counters because we are called again
+      cancel_restart_timers
+
+      # backoff restarts
+      backoff = @restarts ** 2
+      backoff = MAX_RESTART_BACKOFF if backoff > MAX_RESTART_BACKOFF
```

Comment:
yes, and it should be able to differentiate sync-loop vs. actual triggered deploy also. sounds like too many flags for little benefits. :)